### PR TITLE
feat: add support for private registry

### LINKED ISSUE #271: Allow for pulling customHyperkubeImage from private repository
**Is this a request for help?**: 
---
no

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
---
feature request

**What version of acs-engine?**:
---
any

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
kuberentes 

**What happened**:
Would like to be able to pull `customHyperkubeImage` as describe in https://github.com/Azure/acs-engine/blob/master/docs/kubernetes/k8s-developers.md from a private Azure Container Registry. 

Would it be possible to pass Service Principle credentials or use MSI to pull the image?  What would be the effort for making this possible?

Looks like this happens in https://github.com/Azure/acs-engine/blob/8b616722debcddaf0d0d9d2baa605bb7aa77eddb/parts/k8s/kubernetesinstalls.sh#L178:1

**Anything else we need to know**:
cc: @PatrickLang and @ritazh.

### PR DESCRIPTION
**Pull request by ritazh**
**Reason for Change**:

Add support to pull custom hyperkube image from private registry

**Issue Fixed**:

Fixes #271 
This should also fix the broken cloud-provider-azure prow job as that uses ACR.

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

cc @feiskyer

### CODE REVIEW

File: pkg/engine/params_k8s.go [hunk 1]
```diff
@@ -46,6 +46,7 @@ func assignKubernetesParameters(properties *api.Properties, parametersMap params
 
 			addValue(parametersMap, "kubeDNSServiceIP", kubernetesConfig.DNSServiceIP)
 			addValue(parametersMap, "kubernetesHyperkubeSpec", kubernetesHyperkubeSpec)
+			addValue(parametersMap, "privateRegistryServer", kubernetesConfig.PrivateRegistryServer)
```
Comment (jackfrancis) [inline pkg/engine/params_k8s.go]: Should we only add the param if it's not empty string?

File: parts/k8s/kubernetesinstalls.sh [hunk 2]
```diff
@@ -273,6 +273,11 @@ installKubeletAndKubectl() {
 pullContainerImage() {
     CLI_TOOL=$1
     DOCKER_IMAGE_URL=$2
+    if [ ! -z "${PRIVATE_REGISTRY_SERVER}" ]; then
+        if [[ "$CLI_TOOL" == "docker" ]]; then
+            $CLI_TOOL login -u $SERVICE_PRINCIPAL_CLIENT_ID -p $SERVICE_PRINCIPAL_CLIENT_SECRET $PRIVATE_REGISTRY_SERVER
```
Comment (jackfrancis) [inline parts/k8s/kubernetesinstalls.sh]: Does this in practice not work w/ `img`?

Comment (jackfrancis) [reply parts/k8s/kubernetesinstalls.sh]: Seems like it should:

`https://github.com/genuinetools/img`

Comment (jackfrancis) [reply parts/k8s/kubernetesinstalls.sh]: In any event, if `img login` doesn't work in this context, we'll want to document something like the following:

"Private registry support is for default docker (moby) runtime only. It is not supported for non-docker cluster configuration scenarios, e.g., containerd, clear-containers, kata-containers."

Comment (ritazh) [reply parts/k8s/kubernetesinstalls.sh]: Added support for `img` as well.

File: docs/topics/kubernetes-developers.md [hunk 3]
```diff
@@ -43,6 +43,16 @@ For VERSION, we recommend that you provide a value which would help you identify
 }
 ```
 
+* If the container registry is private, for example Azure Container Registry, then provide the name of the private registry along with the custom hyperkube image like this:
+
+```
+"kubernetesConfig": {
+    "customHyperkubeImage": "<your-private-registry>/hyperkube-amd64:<your-custom-version>",
+    "privateRegistryServer": "<your-private-registry>"
```
Comment (jackfrancis) [inline docs/topics/kubernetes-developers.md]: Does privateRegistryServer make any sense in a context when customHyperkubeImage is not provided? If not, we should probably add validation.

**Comment by jackfrancis** [comment]
Final high-level comment: because `privateRegistryServer` seems to be implemented in a MS-specific context (i.e., service principals are accepted auth), should be call this `privateMSRegistryServer`, or something like that? I don't see this implementation working with any old private image repository, or am I missing something?

**Comment by ritazh** [comment]
@jackfrancis In theory this code can be expanded to work for other private container repositories as well, but the real reason is that only azure credentials can actually pull this image from ACR if you are using the Kubernetes Azure cloud provider. So I will update this variable to reflect that.

Should we change it to `privateMSRegistryServer` or `privateAzureRegistryServer`?

**Comment by jackfrancis** [comment]
How about just `azureRegistryServer`?

**Comment by ritazh** [comment]
Can we keep `private` as I think there are plans to enable public ACR in the future? WDYT?

File: pkg/api/vlabs/validate.go [hunk 4]
```diff
@@ -1122,6 +1122,28 @@ func (k *KubernetesConfig) Validate(k8sVersion string, hasWindows bool) error {
...
+
+	// Check PrivateAzureRegistryServer has a valid value.
+	valid := false
+	if k.PrivateAzureRegistryServer != "" {
+		if k.CustomHyperkubeImage != "" {
+			valid = true
+		}
+	} else {
+		valid = true
+	}
```
Comment (jsturtevant) [inline pkg/api/vlabs/validate.go]: I found this logic hard to read.  would something like the following be more more readable?  I think that would remove the need for the `valid` check

```go
k.PrivateAzureRegistryServer != "" and k.CustomHyperkubeImage == "" {
    return errors.Errorf("customHyperkubeImage must be provided when privateAzureRegistryServer is provided")
}
```

File: parts/k8s/kubernetesinstalls.sh [hunk 5]
```diff
@@ -278,6 +278,9 @@ installKubeletAndKubectl() {
 pullContainerImage() {
     CLI_TOOL=$1
     DOCKER_IMAGE_URL=$2
+    if [ ! -z "${PRIVATE_AZURE_REGISTRY_SERVER}" ]; then
+        $CLI_TOOL login -u $SERVICE_PRINCIPAL_CLIENT_ID -p $SERVICE_PRINCIPAL_CLIENT_SECRET $PRIVATE_AZURE_REGISTRY_SERVER
```
Comment (jsturtevant) [inline parts/k8s/kubernetesinstalls.sh]: should this use `--password-stdin` here?  https://docs.docker.com/engine/reference/commandline/login/#provide-a-password-using-stdin

File: parts/k8s/kubernetesmastervars.t [hunk 6]
```diff
@@ -145,7 +145,7 @@
...
...g').cloudProviderBackoffDuration,' CLOUDPROVIDER_BACKOFF_JITTER=',parameters('cloudproviderConfig').cloudProviderBackoffJitter,' CLOUDPROVIDER_RATELIMIT=',toLower(string(parameters('cloudproviderConfig').cloudProviderRatelimit)),' CLOUDPROVIDER_RATELIMIT_QPS=',parameters('cloudproviderConfig').cloudProviderRatelimitQPS,' CLOUDPROVIDER_RATELIMIT_BUCKET=',parameters('cloudproviderConfig').cloudProviderRatelimitBucket,' USE_MANAGED_IDENTITY_EXTENSION=',variables('useManagedIdentityExtension'),' USER_ASSIGNED_IDENTITY_ID=',variables('userAssignedClientID'),' USE_INSTANCE_METADATA=',variables('useInstanceMetadata'),' LOAD_BALANCER_SKU=',variables('loadBalancerSku'),' EXCLUDE_MASTER_FROM_STANDARD_LB=',variables('excludeMasterFromStandardLB'),' MAXIMUM_LOADBALANCER_RULE_COUNT=',variables('maximumLoadBalancerRuleCount'),' CONTAINER_RUNTIME=',parameters('containerRuntime'),' CONTAINERD_DOWNLOAD_URL_BASE=',parameters('containerdDownloadURLBase'),' POD_INFRA_CONTAINER_SPEC=',parameters('kubernetesPodInfraContainerSpec'),' KMS_PROVIDER_VAULT_NAME=',variables('clusterKeyVaultName'),' IS_HOSTED_MASTER={{IsHostedMaster}} PRIVATE_AZURE_REGISTRY_SERVER=',parameters('privateAzureRegistryServer'))]",
```
Comment (tariq1890) [inline parts/k8s/kubernetesmastervars.t]: This should have been separated by a comma. I'll fix this in another PR.

**Pull request merged**